### PR TITLE
DSEGOG-279 data export

### PR DESCRIPTION
**Pull request by kevinphippsstfc**
This is the bulk of the export functionality which will later need to be extended to include the export of function channels when the functions PR has been merged.

The export endpoint mirrors the format of the records endpoint with the addition of a few options allowing control over whether, for example, images of waveforms are included (in addition to CSV files containing the waveform values).

The response will be either a CSV or zip file for download. Using Postman the CSV file can be seen in the response body tab but for zip files you need to use the "Save Response" button to save the zip file to disk.

### CODE REVIEW

**Comment by MRichards99** [comment]
I've just merged main into this branch (which includes the waveform Echo changes) and fixed the path given to `Waveform.get_waveform()` when downloading waveforms for the export functionality as this was causing 3 tests to fail. I'll have a look at reviewing at some point :)

File: operationsgateway_api/src/records/export_handler.py [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+            # so return it
+            return dict_or_value
+        else:
+            # we still have a dictionary which needs further processing
+            # to get down to lower levels where the value will be
+            new_projection = projection.split(".", 1)[1]
+            return ExportHandler.get_value_from_dict(dict_or_value, new_projection)
+
+    @staticmethod
+    def check_zip_file_size(zip_file_in_memory: io.BytesIO) -> None:
```
Comment (MRichards99) [inline operationsgateway_api/src/records/export_handler.py:417]: All the code that calls this function are within this class and none of the functions where the calls occur are static methods. The value that's passed in is always part of `self` too. Does this function need to be static?

File: operationsgateway_api/src/records/export_handler.py [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+                len(self.zip_file.infolist()) > 0
+                and len(self.main_csv_file_in_memory.getvalue()) > 0
+            ):
+                self.zip_file.writestr(
+                    f"{self.get_filename_stem()}.csv",
+                    self.main_csv_file_in_memory.getvalue(),
+                )
+                ExportHandler.check_zip_file_size(self.zip_file_in_memory)
+
+    def _add_messages_file_to_zip(self):
```
Comment (MRichards99) [inline operationsgateway_api/src/records/export_handler.py:312]: Could some logging be added to log out what errors are being written to the file please? When I was fixing the failing test due to merging the echo waveforms changes in, the output of the test mentioned the `EXPORT_ERRORS.txt` file. But as I didn't have access to the zip (as a service manager might not in production), it was difficult to work out what the errors were.

**Review by MRichards99 — CHANGES_REQUESTED** [review]
The functionality works well and the query parameters on the Postman collection made it nice and easy to test, sensible values provided. I've left some comments, some things might not be relevant to how the functionality will actually work.

File: test/conftest.py [hunk 3]
```diff
@@ -71,6 +72,15 @@ def assert_thumbnails(record: dict, expected_thumbnails_hashes: dict):
     assert num_channels_found == len(expected_thumbnails_hashes.keys())
 
 
+def assert_text_file_contents(filepath: str, file_contents: str):
```
[the remark this answers is not in the document]
Comment (kevinphippsstfc) [reply test/conftest.py]: Fixed in https://github.com/ral-facilities/operationsgateway-api/pull/104/commits/45a988a5109355701a4bba9622702594c5275205

File: operationsgateway_api/src/records/export_handler.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+        # if none of the false colour parameters are set then the original
+        # image is required
+        original_image = (
+            self.lower_level == 0
+            and self.upper_level == 255
+            and self.colourmap_name is None
+        )
+        try:
+            if record_id == "20220407142816":
+                raise ExportError("Couldn't find this file!")
```
[the remark this answers is not in the document]
Comment (kevinphippsstfc) [reply operationsgateway_api/src/records/export_handler.py]: Good spot! Fixed in https://github.com/ral-facilities/operationsgateway-api/pull/104/commits/d7ab9718e1dd56e73449c90aa2cb99b8b3e9cfda

File: operationsgateway_api/src/records/export_handler.py [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+        for record_data in self.records_data:
+            log.info("record_data: %s", record_data)
+            record_id = record_data["_id"]
+            self.record_ids.append(record_id)
+            log.info("record_id: %s", record_id)
+
+            line = ""
+            for proj in self.projection:
+                projection_parts = proj.split(".")
+                channel_name = projection_parts[1]
```
[the remark this answers is not in the document]
Comment (kevinphippsstfc) [reply operationsgateway_api/src/records/export_handler.py]: As per my other comment on exporting _id

File: operationsgateway_api/src/records/export_handler.py [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+        record_ids_sorted = sorted(record_ids)
+        first = record_ids_sorted[0]
+        if len(record_ids_sorted) == 1:
+            last = None
+        else:
+            last = record_ids_sorted[-1]
+        return first, last
+
+    @staticmethod
+    def add_value_to_csv_line(line, value) -> str:
```
[the remark this answers is not in the document]
Comment (kevinphippsstfc) [reply operationsgateway_api/src/records/export_handler.py]: Fixed in https://github.com/ral-facilities/operationsgateway-api/pull/104/commits/4c78348dd2c3af14979696f5d71b0f278e3d1b3d

File: operationsgateway_api/src/records/export_handler.py [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+        elif self.channel_manifest_dict["channels"][channel_name]["type"] == "waveform":
+            log.info("Channel %s is a waveform", channel_name)
+            try:
+                x_units = record_data["channels"][channel_name]["metadata"]["x_units"]
+            except KeyError:
+                x_units = ""
+            try:
+                y_units = record_data["channels"][channel_name]["metadata"]["y_units"]
+            except KeyError:
+                y_units = ""
```
[the remark this answers is not in the document]
Comment (kevinphippsstfc) [reply operationsgateway_api/src/records/export_handler.py]: Thanks for the tip! Fixed as suggested in https://github.com/ral-facilities/operationsgateway-api/pull/104/commits/f8ada35c0ac260522b924cf954a0a69393be432b

File: operationsgateway_api/src/records/export_handler.py (hunk 2 above)
Comment (kevinphippsstfc) [reply operationsgateway_api/src/records/export_handler.py:322]: Good idea! Fixed in https://github.com/ral-facilities/operationsgateway-api/pull/104/commits/c071491d22a949d773ff3ea965b7c5c632bc75e8

**Review by MRichards99 — APPROVED** [review]

**Pull request merged**